core/src/main/java/com/linecorp/armeria/client/DefaultDnsQueryLifecycleObserver.java

```diff
@@ -161,26 +160,18 @@ public void querySucceed() {
     private static DnsExceptionTypes determineDnsExceptionTag(Throwable cause) {
         if (cause instanceof DnsTimeoutException) {
             return DnsExceptionTypes.SERVER_TIMEOUT;
-        }
-
-        if (cause instanceof DnsNameResolverTimeoutException) {
+        } else if (cause instanceof DnsNameResolverTimeoutException) {
             return DnsExceptionTypes.RESOLVER_TIMEOUT;
         }
 
         final String message = cause.getMessage();
```

Comment:
Checked `queryFailed()` usages in `DnsResolveContext`, no chance to returns null.
- https://github.com/netty/netty/blob/4.1/resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java#L462
- https://github.com/netty/netty/blob/4.1/resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java#L595-L604
- ...

Reply:
I said "no chance" as `IOException` seemed to have the message, but it is wrong. In many cases, `IOException` has null message. And non-deterministic as you said. Even if not, it would be better to consider null for safety. Updated. PTAL 🙏